components/user-mgt/org.wso2.carbon.identity.user.pre.update.profile.action/src/main/java/org/wso2/carbon/identity/user/pre/update/profile/action/internal/execution/PreUpdateProfileRequestBuilder.java

```diff
@@ -0,0 +1,327 @@
+        for (Map.Entry<String, String> claimEntry : claimValues.entrySet()) {
+            String claimKey = claimEntry.getKey();
+            String claimValue = claimEntry.getValue();
+
+            if (isRoleOrGroupClaim(claimKey) || StringUtils.isBlank(claimValue)) {
+                continue;
+            }
+
+            if (isMultiValuedClaim(claimKey)) {
+                userClaimValuesToSetInEvent.add(
+                        constructMultiValuedClaim(updatingUserClaimsInRequest, claimKey, claimValue,
+                                multiAttributeSeparator));
+            } else {
+                userClaimValuesToSetInEvent.add(constructSingleValuedClaim(updatingUserClaimsInRequest, claimKey,
+                        claimValue));
+            }
+        }
+
+        userBuilder.claims(userClaimValuesToSetInEvent);
+    }
+
+    private void setGroupsInUserBuilder(User.Builder userBuilder, Map<String, String> claimValues,
+                                        String multiAttributeSeparator) {
+
+        if (claimValues.get(GROUP_CLAIM_URI) != null) {
+            userBuilder.groups(Arrays.asList(
+                    claimValues.get(GROUP_CLAIM_URI).split(Pattern.quote(multiAttributeSeparator))));
+        }
+    }
+
+    private boolean isRoleOrGroupClaim(String claimKey) {
+
+        return ROLE_CLAIM_URI.equals(claimKey) || GROUP_CLAIM_URI.equals(claimKey);
+    }
+
+    private UpdatingUserClaim constructMultiValuedClaim(Map<String, Object> updatingUserClaimsInRequest,
+                                                        String claimKey, String claimValue,
+                                                        String multiAttributeSeparator)
+            throws ActionExecutionRequestBuilderException {
+
+        if (updatingUserClaimsInRequest.containsKey(claimKey)) {
+            Object updatingClaimValue = updatingUserClaimsInRequest.get(claimKey);
+            if (!(updatingClaimValue instanceof String[])) {
+                throw new ActionExecutionRequestBuilderException(
+                        "Invalid claim value format for multi-valued claim: " + claimKey +
+                                " Only String[] types are expected.");
+            }
+            return new UpdatingUserClaim(claimKey, claimValue.split(Pattern.quote(multiAttributeSeparator)),
+                    (String[]) updatingClaimValue);
+        }
+
+        return new UpdatingUserClaim(claimKey,
+                claimValue.split(Pattern.quote(multiAttributeSeparator)));
+    }
+
+    private UpdatingUserClaim constructSingleValuedClaim(Map<String, Object> updatingUserClaimsInRequest,
+                                                         String claimKey, String claimValue) {
+
+        if (updatingUserClaimsInRequest.containsKey(claimKey)) {
+            Object updatingClaimValue = updatingUserClaimsInRequest.get(claimKey);
+            return new UpdatingUserClaim(claimKey, claimValue, String.valueOf(updatingClaimValue));
```

Comment:
Shouldn't we check whether updatingClaimValue is string type?

Reply:
This claim map is retrieved from UserActionRequestDTO which gates that only String or String[] types could be set.
If any new type gets onboarded that will need to introduce a new setter and and it's usages anyway needs to be checked explicitly.